### DISCUSSION TITLE
Multivariate models: conditioning on future values?

### DISCUSSION THREAD

**Discussion by timoschowski** [awslabs/gluonts#1379] [Q&A]
Let's say, I have a trained, multi-variate model which allows me to forecast a 2-dimensional time series into the future: 

```
from gluonts.dataset.multivariate_grouper import MultivariateGrouper
from gluonts.dataset.common import TrainDatasets

target_dim = 2
grouper_train = MultivariateGrouper(max_target_dim=target_dim)
grouper_test = MultivariateGrouper(max_target_dim=target_dim)
ds = TrainDatasets(
         metadata="",
        train=grouper_train(training_data),
        test=grouper_test(test_data),
    )
```

and then

```
from gluonts.model.deepvar import DeepVAREstimator

estimator = DeepVAREstimator(freq="5min", prediction_length=200, trainer=Trainer(epochs=10),  target_dim=target_dim)
predictor = estimator.train(training_data=ds.train)
```

What I'd like to do now is to provide future values of one of the two dimensions and get the forecast of the other dimension _conditioned on the provided values_. How would I go about doing this? 

```
from gluonts.evaluation.backtest import make_evaluation_predictions

forecast_it, ts_it = make_evaluation_predictions(
    dataset=ds.test, # here I would like to provide some values on which I can condition the other predictions on
    predictor=predictor,
    num_samples=10,
)
```

Thanks for the help in advance!

**Comment by benidis** [comment]
I can see 2 ways to achieve this.

**1. Treat the known time series as a feature**  
Assume a general case where you have N time series out of which `N1` are known into the future and you want to condition on their values and `N2` are not known, with `N=N1+N2`. Then treat the N1 time series as time features, which are eitherway assumed to be known into the future, and the `N2` as a multivariate time series of dimension `N2`. Then, w.l.o.g. assuming a discriminative model,  the model will learn the conditional distribution based on the lags AND the time features.

This is the most straightforward and less painful way and most of the models should work right away with no changes. 

**2. You can compute the conditional from the joint**
This can work for models that give a `DistributionForecast` output and not a `SampleForecast` output. In this case you can add a method into the model that computes the conditional (in closed form for example if it is Gaussian or using some integral approximations) and then takes samples or whatever you want to do from this instead of the joint. Basically you need to override the sampling of the model to produce samples only from the conditional and then maybe append the known values so your output has the same dimension. 

In a model like NKF you can treat this as missing values as well and there is closed form solution to this.

**Comment by timoschowski** [comment]
Thanks for the answer @benidis -- my example wasn't well chosen, because I'm interested in the multi-variate output distribution  which DeepVAR wouldn't give me. So, `gpvar` would have been the better example. 

If I choose `gpvar` then the pragmatic approach (1) doesn't hold anymore w.l.o.g., I think. 

I was thinking for (2) to instead fit a [multi-variate Gaussian](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.multivariate_normal.html) to the sample paths and then condition on this as a small work around, do you think that that makes sense?

**Comment by benidis** [comment]
I guess it depends on the actual problem you are dealing with. In the general case this cannot work since the number of samples you will need grows exponentially with the number of dimensions. If you actually have a 2D problem then this could be a workaround I guess. However, using the imputation mode of NKF (not sure what is the status of the current version) does exactly what you want and it should be fine regardless of dimension.

**Reply by timoschowski** [reply]
Thanks. To summarize a verbal discussion: for NKF this wouldn't yet work because partially missing values aren't currently supported.

As I'm looking at constructing toy examples, I'll try my work around and hope I'll escape the curse of dimensionality.
